src/careamics/dataset_ng/dataset/dataset.py

```diff
@@ -0,0 +1,154 @@
+    PatchSpecsGenerator,
+    RandomPatchSpecsGenerator,
+    SequentialPatchSpecsGenerator,
+)
+from careamics.transforms import Compose
+
+
+class ImageRegionData(NamedTuple):
+    data: NDArray
+    filename: Path | Literal["array"]
+    data_shape: Sequence[int]
+    dtype: str  # dtype should be str for collate
+    axes: str
+    # TODO: what to do in case of inference on the whole image?
+    #  patch spec of whole image or none?
+    region_spec: PatchSpecs
+
+
+# TODO: how to handle custom data types?
+InputType = Sequence[np.ndarray] | Sequence[Path]
+
+
+class CareamicsDataset(Dataset):
+    # TODO: dataset train / val splitting, should it be in an upper layer?
+    # TODO: if validation is initialized separately, how to disable transforms?
+    # TODO: create file description
+    def __init__(
+        self,
+        data_config: DataConfig | InferenceConfig,
+        inputs: InputType,
+        targets: Optional[InputType] = None,
+        read_func: Optional[Callable] = None,
+    ):
+        self.config = data_config
+
+        self.input_extractor = create_patch_extractor(
+            self.config, data=inputs, read_func=read_func
+        )
+        if targets is not None:
+            self.target_extractor: Optional[PatchExtractor] = create_patch_extractor(
+                self.config, data=targets, read_func=read_func
+            )
+        else:
+            self.target_extractor = None
+
+        self.patch_specs = self._initialize_patch_specs()
+
+        self.input_stats, self.target_stats = self._initialize_statistics()
+
+        self.transforms = self._initialize_transforms()
+
+    def _initialize_patch_specs(self) -> list[PatchSpecs]:
+        if isinstance(self.config, DataConfig):
+            # TODO: how to fix the random seed properly?
+            #  how to change it between epochs?
+            patch_generator: PatchSpecsGenerator = RandomPatchSpecsGenerator(
+                patch_size=self.config.patch_size,
+                random_seed=getattr(self.config, "random_seed", 42),
+            )
+        elif isinstance(self.config, InferenceConfig):
+            # TODO: how to handle the whole images?
+            if self.config.tile_size is None or self.config.tile_overlap is None:
+                raise ValueError(
+                    "InferenceConfig must specify tile_size and tile_overlap"
+                )
+            patch_generator = SequentialPatchSpecsGenerator(
+                patch_size=self.config.tile_size, overlap=self.config.tile_overlap
+            )
+        else:
+            raise ValueError(f"Data config type {type(self.config)} is not supported.")
+
+        patch_specs = patch_generator.generate(data_shapes=self.input_extractor.shape)
+        return patch_specs
+
+    def _initialize_transforms(self) -> Optional[Compose]:
+        if isinstance(self.config, DataConfig):
+            return Compose(
+                transform_list=list(self.config.transforms),
```

Comment:
Config currently does not include normalization, since normalization is always applied.

We could add `no_normalization` parameter to the configuration and an `if` clause here to add it or not to the list before the compose.

Reply:
No need to do it in this PR, but it should be a TODO and an issue (if not there already).